presto-test-jdbc-compatibility-old-server/src/test/java/io/prestosql/TestJdbcResultSetCompatibility.java

```diff
@@ -0,0 +1,125 @@
+        extends BaseTestJdbcResultSet
+{
+    private static final int NUMBER_OF_TESTED_VERSIONS = 5;
+    private static final int TESTED_VERSIONS_GRANULARITY = 3;
+
+    private final Optional<String> testedPrestoVersion;
+    private PrestoContainer<?> prestoContainer;
+
+    @Factory(dataProvider = "testedPrestoVersions")
+    public TestJdbcResultSetCompatibility(Optional<String> testedPrestoVersion)
+    {
+        this.testedPrestoVersion = requireNonNull(testedPrestoVersion, "testedPrestoVersion is null");
+    }
+
+    @DataProvider
+    public static Object[][] testedPrestoVersions()
+    {
+        try {
+            String currentVersionString = PrestoDriver.class.getPackage().getImplementationVersion();
+            if (currentVersionString == null) {
+                return new Object[][] {
+                        {Optional.empty()}
+                };
+            }
+
+            int currentVersion = Integer.parseInt(Splitter.on("-").omitEmptyStrings().trimResults().splitToList(currentVersionString).get(0));
+            ImmutableList.Builder<String> testedPrestoVersions = ImmutableList.builder();
+            int testedVersion = currentVersion - 1;
+            for (int i = 0; i < NUMBER_OF_TESTED_VERSIONS; i++) {
+                testedPrestoVersions.add(String.valueOf(testedVersion));
+                testedVersion -= TESTED_VERSIONS_GRANULARITY;
+            }
+
+            return testedPrestoVersions.build().stream()
+                    .map(version -> new Object[] {Optional.of(version)})
+                    .toArray(Object[][]::new);
+        }
+        catch (Throwable e) {
+            System.err.println("Could not determine Presto versions to test; " + e.getMessage() + "\n" + getStackTrace(e));
+            return new Object[][] {
+                    {Optional.empty()}
+            };
+        }
+    }
+
+    private static String getStackTrace(Throwable e)
+    {
+        StringWriter stringWriter = new StringWriter();
+        PrintWriter printWriter = new PrintWriter(stringWriter);
+        e.printStackTrace(printWriter);
+        return stringWriter.toString();
+    }
+
+    @BeforeClass
+    public void setupPrestoContainer()
+    {
+        if (testedPrestoVersion.isEmpty()) {
+            throw new AssertionError("Could not determine current Presto version");
+        }
+        Logging.initialize();
+
+        prestoContainer = new PrestoContainer<>("prestosql/presto:" + testedPrestoVersion.get());
+        prestoContainer.start();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDownPrestoContainer()
+    {
+        if (prestoContainer != null) {
+            prestoContainer.stop();
+            prestoContainer = null;
+        }
+    }
+
+    @Override
+    protected Connection createConnection()
+            throws Exception
+    {
+        return DriverManager.getConnection(prestoContainer.getJdbcUrl(), "test", null);
+    }
+
+    @Override
+    public String toString()
```

Comment:
Is it used by Testng? how?

(please add a comment; we do not use `@Factory` tests elsewhere so it's not a common knowledge currently)